### PR TITLE
[stdlib] Simplify concrete floating-point-to-integer conversion

### PR DESCRIPTION
**Pull request by xwu**
This PR is a minor simplification of concrete floating-point conversions to integers. Specifically:

* It preserves the GYB logic improved in #16960 that more robustly checks when FP-to-integer conversion is always exact for two given types, but refactors the logic in one place so that subsequent tests can use a single Boolean value `is_always_represented_exactly`.

* It restores the use of roundtripping as the test for exactness, which is the notional litmus test. This will avoid the scenario where any future changes could cause roundtripping and manually implemented tests of exactness to fall out of sync, it eliminates an untestable branch without changing behavior, and it improves ease of understanding of the code particularly now that the function bodies are greatly simplified.

For FP-to-integer conversions, this PR preserves the `guard` condition implemented in #16960 that prevents undefined results, but corrects the comment to note that it is the initial guard condition that guards against NaN source values, and expands on that comment to explain that we are guarding against undefined results.

### CODE REVIEW

**Comment by xwu** [comment]
@swift-ci Please smoke test

File: stdlib/public/core/FloatingPointTypes.swift.gyb [hunk 1]
```diff
@@ -1589,15 +1589,14 @@ extension ${Self} {
...
-%   maxBitLength = SignificandBitCount + 1 # implicit leading one
-%   if self_ty.is_signed:
-%     # The magnitude of the minimum value of a signed integer type is always
-%     # representable in floating point (assuming a large enough exponent)
-%     # because only one bit is set. All other possible magnitudes can be
-%     # represented using one bit less than the original source type.
-%     maxBitLength += 1
-%   end
+%
+%   max_required_exponent = srcBits - 1
```
Comment (mundaym) [inline stdlib/public/core/FloatingPointTypes.swift.gyb]: All the other names use camel case. Is there a reason to switch? I don't have a preference (if anything I prefer using underscores) but consistency makes things easier to read.

File: stdlib/public/core/Integers.swift.gyb [hunk 2]
```diff
@@ -3291,15 +3291,16 @@ public struct ${Self}
...
   public init?(exactly source: ${FloatType}) {
+    // The value passed as `source` must not be infinite, NaN, or exceed the
+    // bounds of the integer type; the result of `fptosi` or `fptoui` is
+    // undefined if it overflows.
     guard source > ${str(lower)}.0 && source < ${str(upper)}.0 else {
-      // The source is out of bounds (including infinities).
       return nil
     }
-    guard source == source.rounded(.towardZero) else {
-      // The source is a fraction or NaN.
```
Comment (mundaym) [inline stdlib/public/core/Integers.swift.gyb]: Good catch.

File: stdlib/public/core/Integers.swift.gyb [hunk 3]
```diff
@@ -3291,15 +3291,16 @@ public struct ${Self}
...
+    self._value = Builtin.fpto${u}i_FPIEEE${FloatBits}_${BuiltinName}(source._value)
+    if ${FloatType}(self) != source {
```
Comment (mundaym) [inline stdlib/public/core/Integers.swift.gyb]: I don't think this code change is an improvement. The `source.rounded(.towardZero)` comparison is clearer and more self contained than the round trip in my opinion.

**Review by mundaym — COMMENTED** [review]
Personally I think if we are worried about the implementations of exact conversions becoming out of sync, and therefore round trip behavior being incorrect, we should add tests for it rather than relying on the implementation doing a round trip internally.

Round trips can be hard to reason about, especially when they don't call the exact constructor. Self contained implementations of conversions can be reasoned about more easily and can potentially be considerably faster.

File: stdlib/public/core/FloatingPointTypes.swift.gyb (hunk 1 above)
Comment (xwu) [reply stdlib/public/core/FloatingPointTypes.swift.gyb]: Not all the other names use camel case, and from what I can glean, there is an actual method to the madness:

Variables that are emitted in some way in the generated Swift source adopt Swift naming conventions which reflect their primary use case in the generated source. In other words, those Python variables that represent types (`ThatBuiltinName`) are capitalized as though a Swift type, and those that represent instance properties (`srcBits`) adopt camelCase.

Notice, however, that other variables defined in Python (`self_ty.is_signed`) use snake_case. Here, none of these variables are emitted into the Swift source code; therefore, after some initial waffling, I have made them snake_case.

**Comment by xwu** [comment]
The principal impetus for my creating this PR is in fact to address the issue of manual implementations producing unanticipated results with future additions or changes. I have been spending some spare time studying and documenting the behavior of numeric conversions in Swift, and it is (to put it mildly) a bewildering patchwork. I have identified manually implemented conversions being a major root cause of this issue.

At least two issues arise with manual implementations--

__First:__ Because no two types can represent exactly the same set of values, there's always bound to be some user who finds a conversion to be a little more strict or lenient than they personally expect _regardless of how sound the implementation actually is_; bugs get filed, tweaks get made to manual implementations to address the specific example without reference to any overarching guiding semantics, only to open up more problems that weren't there before.

This is not empty theorizing. [That bug](https://bugs.swift.org/browse/SR-6322) you tried to fix in #16960? That's exactly what happened:
1. A bug was filed that conversion from `Double` to `Float` was too strict via `NSNumber`--it failed on loss of precision: `(Double.pi as! NSNumber) as? Float == nil`.
2. A discussion was had on the forums, and with consensus [conversion was made more lenient](https://github.com/apple/swift/commit/c358afe6555e5e32633e879f96a3664dc7a5f3dc#diff-390bd9aed62915ecd0c43c8d6ecf0e08).
3. That was the cause of the bug you intended to fix.
4. How was it solved? Well, [conversions were made more strict again](https://github.com/apple/swift/commit/956e793ef0814c939ef150536e5d207914eefc91#diff-390bd9aed62915ecd0c43c8d6ecf0e08).

We can all anticipate what'll happen some years down the road. Someone will file a bug again that conversion is too strict, it will seem like a good idea to adjust the algorithm, etc. This cycle of bugginess is avoided when the actual implementation of the algorithm directly recites the intended semantics (i.e., roundtripping), leaving no room for tweaks without explicitly reconsidering the overall design.

__Second:__ It's sometimes difficult and _often impossible_ to test such an implementation in a thorough or systematic way (since a systematic approach is required to avoid ad-hoc tweaks to "correct" specific results as described above). It is out of the question to iterate through every possible value of `Int64`, for example, and in the case of conversions we are faced with a combinatorial explosion of possibilities (architecture × OS × source type × destination type, especially daunting for floating-point types because architecture and OS mix in some interesting ways to influence the final behavior).

You may say that it's entirely overkill to iterate through every possible value, but let me show you something:

```swift
import Foundation

(UInt.max as! NSNumber).boolValue
// true
(UInt.max - 1 as! NSNumber).boolValue
// true
// ...
let x = 9223372036854775810 as UInt64
(x as! NSNumber).boolValue
// true
(x - 1 as! NSNumber).boolValue
// true
(x - 2 as! NSNumber).boolValue
// false (!)
(x - 3 as! NSNumber).boolValue
// true
(x - 4 as! NSNumber).boolValue
// true
// ...
```

We've got a rather awkward bug in conversion to a destination type with only _two possible values_! (This is a known bug, and the behavior is preserved now for compatibility. For fun, see if you can work out why it happens; if you're stuck, it's documented in an obscure comment in the swift-corelibs-foundation implementation of `NSNumber` or read on for spoilers.) Could this have been picked up without testing every value? Sure. But it involves testing the bounds of __`Int64`__ in a case of __`UInt64`__-to-`Bool` conversion, hardly an obvious choice.

And I'm not merely cherry-picking one issue with `NSNumber` conversions implemented manually. Their behavior has changed with virtually every recent major and minor release of Swift. Bugs are still being found today (I believe @natecook1000 just made some adjustments this week), and some of the surprising behavior doesn't even arise from bugs in implementation but rather are fundamental consequences of the chosen design:

```swift
import Foundation

(Double(Int32.max - 1) as! NSNumber) == (Int32.max as! NSNumber)
// false
(Double(Int64.max - 1) as! NSNumber) == (Int64.max as! NSNumber)
// true
// This result will become `false` if `Int128` becomes supported.
```

Although it isn't the case with the previous example, when it comes to future-proofing the implementation, an appropriate test can also sometimes be _impossible_ to write without the future feature itself even if you can anticipate what that feature might be. Consider the issue with possible future integer types that can represent values exceeding the max supported exponent of `Float`: the code paths in your manual implementation to handle that conversion _cannot_ be tested today.

__Therefore,__ the best solution in my view is to restore actual roundtripping whenever possible, which fortuitously is Swift's preferred semantics for exact conversion: not only is it imminently understandable when explained to users, it is also implementable fairly efficiently. Of course, it is exceedingly important to do what you did in #16960 in guarding against undefined behavior, and if performance truly becomes an issue we can optimize as necessary. But in the absence of other exigencies, adopting the following implementation uniformly allows one to verify that the code has the desired semantics _by inspection_ and prevents the pitfalls delineated above suffered by `NSNumber`: (1) guard against input that produces undefined output; (2) convert once to destination type; (3) check for equality after converting a second time back to the source type.

**Comment by stephentyrone** [comment]
@xwu Round-tripping is not sufficient. It's good enough for conversion between floating-point and integer types, but would not suffice for e.g. conversions between binary and decimal floating-point types, where two values commonly round to each other, but are still not equal.

The concrete conversions should be as efficient as possible, and we should have test cases that exercise the interesting corners, even for types that haven't been defined yet (we know what the rough shape of those types would be, so we can still write tests with some thought).

All that said, I would like to rework and clean-up these conversions more thoroughly in the short-term future anyway (specifically moving as much as possible to the protocol-level while maintaining performance, which will make testing a bit nicer), so I'm OK with taking these changes as-is.

**Comment by xwu** [comment]
> It's good enough for conversion between floating-point and integer types, but would not suffice for e.g. conversions between binary and decimal floating-point types, where two values commonly round to each other, but are still not equal.

I see the outlines of what you're getting at, but I can't identify any actual examples where we wouldn't want them to be equal. What do you have in mind?

**Comment by mundaym** [comment]
@xwu:

> Therefore, the best solution in my view is to restore actual roundtripping whenever possible

To be clear, it's fine to use the `exactly:` constructor in round trips since it is well defined and easy to reason about. I think it's better to do something self contained to avoid having to jump around the code when debugging but this is a minor personal preference. Performing an inexact cast to test the round trip (like`if ${FloatType}(self) != source`) isn't fine though. In this case it replaces an operation with clearly defined behavior with one that has undefined behavior in some situations (albeit guarded against earlier in the code). That doesn't seem like a win.

> It's sometimes difficult and often impossible to test such an implementation in a thorough or systematic way

As @stephentyrone mentioned there are strategies to get good coverage of all edge cases even without testing the whole range of values. I think it would be good to improve the test cases in this way as the better the tests are the less we need to worry about the correctness of the implementation and the more we can focus on readability and speed.

> some of the surprising behavior that we'll be stuck with for a while yet doesn't even arise from bugs in implementation but rather are fundamental consequences of the chosen design:

```swift
(Double(Int64.max - 1) as! NSNumber) == (Int64.max as! NSNumber)
// true
// This result will become `false` if `Int128` becomes supported.
```

Why would this become false? I haven't dug that far into the code, but presumably this condition results in true because the Int64 on the right hand side is promoted to Double at some point which *wouldn't* change if Int128 were added. This may or may not be the desired behavior :).

> Consider the issue with possible future integer types that can represent values exceeding the max supported exponent of Float: the code paths in your manual implementation to handle that conversion cannot be tested today.

I'd be happy for that code path to be taken out and replaced with an assertion. The LLVM conversions the code uses don't exist for non-standard types so it would have to be rewritten if we wanted to support them anyway.

tl;dr: if you revert the code changes to `Integers.swift.gyb` and just fix the comment then I'm happy with this PR.

**Comment by stephentyrone** [comment]
> I see the outlines of what you're getting at, but I can't identify any actual examples where we wouldn't want them to be equal. What do you have in mind?

Here's a simple example. Suppose we have two floating-point formats, a decimal format with one digit D, and a binary format with one explicit significand bit B.

~~~~
let x: D = 0.1 // exact, because D is decimal and this only needs one digit
let y: B = 0.1 // rounds to 0.9375 (1.5 * 2^-4)
let z = D(exactly: y) // nil, since not exactly representable
let w = B(exactly: x) // nil, since not exactly representable
~~~~

**Comment by xwu** [comment]
@stephentyrone That's a good explanation. I hadn't considered cycles of initializers in that scenario (and, in particular, `Float`-to-`Decimal`-to-`Double`-to-`Float` cycles). I can see that round-tripping is necessary but not sufficient to define "exact" equality for those purposes when we have a change in radix (radical change?).

**Review by stephentyrone — APPROVED** [review]

**Comment by stephentyrone** [comment]
@swift-ci please smoke test

**Comment by stephentyrone** [comment]
Thanks for persevering on this one, @xwu!
